src/snmpanon.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" snmpanon is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH SNMPANON 1 "December 2, 2005"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
snmpanon \- anonymize SNMP messages
.SH SYNOPSIS
.B snmpanon
.RI [ options ]
.SH DESCRIPTION
\fBsnmpanon\fP anonymizes SNMP messages. It reads messages from an XML
file (typically created using snmpdump) and applies anonymization to
it. The anonymization is controlled by the option passed to snmpanon.
.PP
Anonymization results can be drastically improved by loading the
relevant MIB modules using the -m option. A default MIB module set can
be specified using the libsmi startup mechanism described in the
smi_config(3) man page.
.SH OPTIONS
\fBsnmpanon\fP accepts the following options:
.TP
\fB-a \fIregex\fB, --anonymize=\fIregex\fP
xxx
.TP
\fB-c \fIxpath\fB, --clear=\fIxpath\fP
Clear all nodes in the XML document matching the \fIxpath\fR
expression by setting empty content. This option can be used multiple
times to match against multiple xpath expression.
.TP
\fB-d \fIxpath\fB, --delete=\fIxpath\fP
Delete all nodes in the XML document matching the \fIxpath\fR.  This
option can be used multiple times to match against multiple xpath
expression.
.TP
\fB-s \fIfile\fB, --smi-config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
libsmi configuration file.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
\fB-m \fIfile\fB \-\-module \fIfile\fB
Load SMI MIB module definitions from \fIfile\fR.  This option can be
used several times to load several MIB modules into snmpanon.
.TP
.B \-V, \-\-version
Show version of program.
.SH "SEE ALSO"
.BR smi_config(3).
.SH AUTHOR
Juergen Schoenwaelder <j.schoenwaelder@iu-bremen.de>.


ChangeLog
2005-06-12 23:49  schoenw

	* Makefile.in, acinclude.m4, aclocal.m4, autogen.sh, config.guess,
	  config.sub, configure, depcomp, doc/Makefile.in, install-sh,
	  missing, mkinstalldirs, src/Makefile.in, test/Makefile.in: Added
	  autogen.sh and removed all autoconf related files from the
	  repository.

2005-06-07 19:57  schoenw

	* AUTHORS, COPYING, ChangeLog, INSTALL, Makefile.am, Makefile.in,
	  NEWS, README, acinclude.m4, aclocal.m4, autom4te.cache,
	  autom4te.cache/output.0, autom4te.cache/requests,
	  autom4te.cache/traces.0, config.guess, config.sub, configure,
	  configure.in, depcomp, doc, doc/Makefile.am, doc/Makefile.in,
	  doc/snmptrace.rnc, install-sh, missing, mkinstalldirs, src,
	  src/Makefile.am, src/Makefile.in, src/config.h.in, src/snmpdump.1,
	  src/snmpdump.c, test, test/Makefile.am, test/Makefile.in,
	  test/frags.pcap, test/misc.pcap, test/scli.pcap, test/traps.pcap:
	  import snmpdump sources

2005-06-07 19:51  root

	* .: snmpdump project trunk



src/snmpanon.c
/*
 * snmpanon.c --
 *
 * A utility to anonymize SNMP messages in XML format (as generated by
 * snmpdump). The anonymized XML output will be written to stdout.
 *
 * (c) 2004-2005 Juergen Schoenwaelder <j.schoenwaelder@iu-bremen.de>
 *
 */

#include "config.h"

#include <string.h>
#include <unistd.h>
#include <inttypes.h>

#include <libxml/xmlmemory.h>
#include <libxml/tree.h>
#include <libxml/xpath.h>

#include <smi.h>

#include "xpath-filter.h"
#include "libanon.h"

static const char *progname = "snmpanon";

static xmlDocPtr xml_doc;

static unsigned char my_key[32] = 
{
     21, 34, 23,141, 51,164,207,128, 19, 10, 91, 22, 73,144,125, 16,
    216,152,143,131,121,121,101, 39, 98, 87, 76, 45, 42,132, 34,  2
};



/*
 *
 */

static void
mark_anon_ip_node(anon_ip_t *an_ip, const char *xpath, xmlXPathContextPtr ctxt)
{
    xmlXPathObjectPtr obj;
    xmlChar *content;
    int i;
    in_addr_t ip;

    obj = xmlXPathEval(BAD_CAST(xpath), ctxt);
    if (obj) {
	if (obj->type == XPATH_NODESET) {
	    for (i = 0; i < xmlXPathNodeSetGetLength(obj->nodesetval); i++) {
		content = xmlNodeGetContent(obj->nodesetval->nodeTab[i]);
		if (inet_pton(AF_INET, (char *) content, &ip) > 0) {
		    anon_ip_set_used(an_ip, ip, 32);
		}
	    }
	}
	xmlXPathFreeObject(obj);
    }
}

static void
repl_anon_ip_node(anon_ip_t *an_ip, const char *xpath, xmlXPathContextPtr ctxt)
{
    xmlXPathObjectPtr obj;
    xmlChar *content;
    int i;
    in_addr_t ip;
    in_addr_t aip;

    obj = xmlXPathEval(BAD_CAST(xpath), ctxt);
    if (obj) {
	if (obj->type == XPATH_NODESET) {
	    for (i = 0; i < xmlXPathNodeSetGetLength(obj->nodesetval); i++) {
		content = xmlNodeGetContent(obj->nodesetval->nodeTab[i]);
		if (inet_pton(AF_INET, (char *) content, &ip) > 0) {
		    char buf[INET_ADDRSTRLEN];
		    (void) anon_ip_map_pref_lex(an_ip, ip, &aip);
		    if (inet_ntop(AF_INET, &aip, buf, sizeof(buf))) {
			xmlNodeSetContent(obj->nodesetval->nodeTab[i],
					  BAD_CAST(buf));
		    }
		}
	    }
	}
	xmlXPathFreeObject(obj);
    }
}

/*
 *
 */

static void
mark_anon_port_node(anon_int64_t *an_ip, const char *xpath, xmlXPathContextPtr ctxt)
{
    xmlXPathObjectPtr obj;
    xmlChar *content;
    int i;
    int64_t num;

    obj = xmlXPathEval(BAD_CAST(xpath), ctxt);
    if (obj) {
	if (obj->type == XPATH_NODESET) {
	    for (i = 0; i < xmlXPathNodeSetGetLength(obj->nodesetval); i++) {
		content = xmlNodeGetContent(obj->nodesetval->nodeTab[i]);
		if (sscanf((char *)content, "%"SCNd64, &num) == 1) {
		    anon_int64_set_used(an_ip, num);
		}
	    }
	}
	xmlXPathFreeObject(obj);
    }
}

static void
repl_anon_port_node(anon_int64_t *an_ip, const char *xpath, xmlXPathContextPtr ctxt)
{
    xmlXPathObjectPtr obj;
    xmlChar *content;
    int i;
    int64_t num, anum;

    obj = xmlXPathEval(BAD_CAST(xpath), ctxt);
    if (obj) {
	if (obj->type == XPATH_NODESET) {
	    for (i = 0; i < xmlXPathNodeSetGetLength(obj->nodesetval); i++) {
		content = xmlNodeGetContent(obj->nodesetval->nodeTab[i]);
                if (sscanf((char *)content, "%"SCNd64, &num) == 1) {
		    char buf[40];
		    (void) anon_int64_map_lex(an_ip, num, &anum);
		    if (snprintf(buf, sizeof(buf), "%"PRId64, anum) > 0) {
			xmlNodeSetContent(obj->nodesetval->nodeTab[i],
					  BAD_CAST(buf));
		    }
		}
	    }
	}
	xmlXPathFreeObject(obj);
    }
}

static void
mark_anon_varbind_name(const char *xpath, xmlXPathContextPtr ctxt)
{
    xmlXPathObjectPtr obj;
    xmlChar *content;
    SmiNode *smiNode;
    SmiType *smiType;
    int i;

    obj = xmlXPathEval(BAD_CAST(xpath), ctxt);
    if (obj) {
	if (obj->type == XPATH_NODESET) {
	    for (i = 0; i < xmlXPathNodeSetGetLength(obj->nodesetval); i++) {
		content = xmlNodeGetContent(obj->nodesetval->nodeTab[i]);
		fprintf(stderr, "** <%s>\n", (char *) content);
		smiNode = smiGetNode(NULL, (char *) content);
		if (smiNode) {
		    fprintf(stdout, "** %s", smiNode->name);
		    smiType = smiGetNodeType(smiNode);
		    if (smiType) {
			fprintf(stdout, " [%s]", smiType->name);
		    }
		    fprintf(stdout, "\n");
		}
	    }
	}
	xmlXPathFreeObject(obj);
    }
}

/*
 * First anonymization transformation pass: collect all the data
 * values that need anonymization and clear the rest.
 */

static void
anon_pass1(anon_ip_t *an_ip, anon_int64_t *an_port)
{
    xmlXPathContextPtr ctxt;

    ctxt = xmlXPathNewContext(xml_doc);
    ctxt->node = xmlDocGetRootElement(xml_doc);

    mark_anon_ip_node(an_ip, "//snmptrace/packet/*/@ip", ctxt);
    mark_anon_port_node(an_port, "//snmptrace/packet/*/@port", ctxt);

    mark_anon_varbind_name("//snmptrace/packet/snmp/*/variable-bindings/varbind/name", ctxt);

    xmlXPathFreeContext(ctxt);
}

/*
 * Second anonymization transformation pass: replace the data
 * looking for anonymization.
 */

static void
anon_pass2(anon_ip_t *an_ip, anon_int64_t *an_port)
{
    xmlXPathContextPtr ctxt;

    ctxt = xmlXPathNewContext(xml_doc);
    ctxt->node = xmlDocGetRootElement(xml_doc);

    repl_anon_ip_node(an_ip, "//snmptrace/packet/*/@ip", ctxt);
    repl_anon_port_node(an_port, "//snmptrace/packet/*/@port", ctxt);
    
    xmlXPathFreeContext(ctxt);
}


int
main(int argc, char **argv)
{
    int i, c;
    char buffer[1024];
    char *filter = NULL;
    xpath_filter_t *xpf;

    xpf = xpath_filter_new();

    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-s") == argv[i]) ||
	    (strstr(argv[i], "--smi-config") == argv[i])) break;
    if (i == argc) {
	smiInit("smilint");
    } else {
	smiInit(NULL);
    }
	
    while ((c = getopt(argc, argv, "Vhc:d:m:")) != -1) {
	switch (c) {
	case 'c':
	    if (xpf) {
		xpath_filter_add(xpf, BAD_CAST(optarg),
				 XPATH_FILTER_TYPE_CLEAR);
	    }
	    break;
	case 'd':
	    if (xpf) {
		xpath_filter_add(xpf, BAD_CAST(optarg),
				 XPATH_FILTER_TYPE_DELETE);
	    }
	    break;
	case 'm':
	    smiLoadModule(optarg);
	    break;
	case 's':
	    smiReadConfig(optarg, "snmpdump");
	    break;
	case 'V':
	    printf("%s %s\n", progname, VERSION);
	    exit(0);
	case 'h':
	case '?':
	    printf("%s [-c xpath] [-d xpath] [-m module] [-h] [-s config] file ... \n", progname);
	    exit(0);
	}
    }

    for (i = 1; i < argc; i++) {
	xml_doc = xmlReadFile(argv[i], NULL, 0);
	if (! xml_doc) {
	    fprintf(stderr, "%s: could not parse XML file '%s'\n",
		    progname, argv[i]);
	    continue;
	}

	/* anonymize the data */
	
	anon_ip_t *an_ip;
	anon_int64_t *an_port;

	xmlXPathInit();

	an_ip = anon_ip_new();
	if (! an_ip) {
	    fprintf(stderr, "%s: initialization of IP anonymization failed\n",
		    progname);
	    exit(1);
	}
	an_port = anon_int64_new(0, 65535);
	if (! an_port) {
	    fprintf(stderr, "%s: initialization of port anonymization failed\n",
		    progname);
	    exit(1);
	}
	
	anon_ip_set_key(an_ip, my_key);
	anon_int64_set_key(an_port, my_key);

	anon_pass1(an_ip, an_port);	/* xml_root passed as global */
	anon_pass2(an_ip, an_port);	/* xml_root passed as global */

	anon_ip_delete(an_ip);
	anon_int64_delete(an_port);

	/* apply the filters */
	
	xpath_filter_apply(xpf, xml_doc);

	/* print the resulting xml document */

	if (xmlDocFormatDump(stdout, xml_doc, 1) == -1) {
	    fprintf(stderr, "%s: failed to serialize xml document\n", progname);
	    exit(1);
	}
    }
    
    /* cleanup */

    xpath_filter_delete(xpf);
}
